File: experimental/touchflow-cpp/src/daemon.cpp
#include "daemon.hpp"
#include "keyboard_view.hpp"

#include <iostream>

namespace touchflow {

Daemon::Daemon()
    : config_(Config::load())
    , learning_(config_.learning_threshold) {}

void Daemon::on_keyboard_action(const std::string& action, const std::string& detail) {
    (void)detail;
    if (action == "hide") hide_keyboard(true);
}

void Daemon::show_keyboard(bool manual) {
    if (config_.hide_on_external_keyboard && ext_kb_ && ext_kb_->connected() && !manual)
        return;
    if (window_) {
        gtk_widget_set_visible(GTK_WIDGET(window_), TRUE);
        gtk_window_present(window_);
        visible_ = true;
        manual_show_ = manual;
        if (manual)
            learning_.on_auto_show(current_focus_.app_id, current_focus_.window_class);
    }
}

void Daemon::hide_keyboard(bool manual) {
    if (window_) {
        learning_.on_dismiss(current_focus_.app_id, current_focus_.window_class, manual);
        gtk_widget_set_visible(GTK_WIDGET(window_), FALSE);
        visible_ = false;
        manual_show_ = false;
    }
}

void Daemon::on_focus(const FocusInfo& info) {
    current_focus_ = info;
    if (!info.is_text_entry) {
        if (config_.auto_hide_on_blur && visible_ && !manual_show_)
            hide_keyboard(false);
        return;
    }
    if (config_.hide_on_external_keyboard && ext_kb_ && ext_kb_->connected())
        return;
    if (!config_.auto_show) return;
    if (!learning_.should_auto_show(info.app_id, info.window_class)) return;
    show_keyboard(false);
    learning_.on_auto_show(info.app_id, info.window_class);
}

void Daemon::on_ext_kb(bool connected, void* data) {
    auto* self = static_cast<Daemon*>(data);
    if (connected && self->config_.hide_on_external_keyboard && self->visible_)
        self->hide_keyboard(false);
}

gboolean Daemon::poll_ext_kb(gpointer data) {
    auto* self = static_cast<Daemon*>(data);
    if (self->ext_kb_) self->ext_kb_->poll();
    return G_SOURCE_CONTINUE;
}

void Daemon::setup_ui(AdwApplication* app) {
    window_ = GTK_WINDOW(adw_application_window_new(GTK_APPLICATION(app)));
    gtk_window_set_title(window_, "TouchFlow C++");
    gtk_window_set_decorated(window_, FALSE);
    gtk_window_set_resizable(window_, FALSE);
    gtk_window_set_default_size(window_, -1, config_.height_px);

    keyboard_ = keyboard_view_new(config_, &injector_,
        [this](const std::string& a, const std::string& d) { on_keyboard_action(a, d); });
    adw_application_window_set_content(ADW_APPLICATION_WINDOW(window_), keyboard_);
}

void Daemon::setup_services() {
    focus_ = std::make_unique<FocusWatcher>([this](const FocusInfo& i) { on_focus(i); });
    focus_->start();
    ext_kb_ = std::make_unique<ExternalKeyboardMonitor>(&Daemon::on_ext_kb, this);
    poll_id_ = g_timeout_add_seconds(2, poll_ext_kb, this);
    if (!injector_.available())
        std::cerr << "touchflow-cpp: key injection disabled — add user to input group\n";
}

int Daemon::run(int argc, char** argv, bool virtual_keyboard) {
    virtual_kb_ = virtual_keyboard;
    auto* app = adw_application_new("com.touchflow.Keyboard.Cpp", G_APPLICATION_DEFAULT_FLAGS);
    g_signal_connect(app, "activate", G_CALLBACK(+[](AdwApplication* app, gpointer data) {
        auto* self = static_cast<Daemon*>(data);
        self->setup_ui(app);
        self->setup_services();
        if (self->virtual_kb_ || !self->config_.startup_hidden)
            self->show_keyboard(self->virtual_kb_);
    }), this);
    int status = g_application_run(G_APPLICATION(app), argc, argv);
    if (poll_id_) g_source_remove(poll_id_);
    g_object_unref(app);
    return status;
}

}  // namespace touchflow

File: touchflow/daemon.py
"""Главный демон TouchFlow."""

from __future__ import annotations

import logging
import sys
import time
from typing import TYPE_CHECKING

import gi

gi.require_version("Gtk", "4.0")
gi.require_version("Gdk", "4.0")
from gi.repository import Gdk, GLib, Gtk

from touchflow import __version__, APP_ID
from touchflow.config import TouchFlowConfig, load_config, save_config
from touchflow.dbus_iface import TouchFlowDBusService
from touchflow.external_kb import ExternalKeyboardMonitor
from touchflow.focus_watcher import FocusInfo, FocusWatcher
from touchflow.gestures import SwipeZone
from touchflow.keyboard_widget import KeyboardWidget
from touchflow.key_inject import KeyInjector
from touchflow.learning import LearningEngine
from touchflow.onboarding import is_first_run, show_onboarding_if_needed
from touchflow.overlay import OverlayWindow
from touchflow.physical_bindings import PhysicalButtonListener

if TYPE_CHECKING:
    pass

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
log = logging.getLogger("touchflowd")


class TouchFlowDaemon:
    def __init__(self, virtual_keyboard: bool = False):
        self._virtual_keyboard = virtual_keyboard
        self.config = load_config()
        self.injector = KeyInjector()
        self.learning = LearningEngine(config=self.config.learning)
        self.kb_monitor = ExternalKeyboardMonitor(on_change=self._on_external_kb_change)
        self.focus_watcher = FocusWatcher(self._on_focus_change)

        self._window: Gtk.Window | None = None
        self._keyboard: KeyboardWidget | None = None
        self._overlay: OverlayWindow | None = None
        self._swipe_zone: SwipeZone | None = None
        self._physical: PhysicalButtonListener | None = None
        self._dbus: TouchFlowDBusService | None = None

        self._visible = False
        self._current_focus = FocusInfo()
        self._show_time = 0.0
        self._manual_show = False

    @property
    def visible(self) -> bool:
        return self._visible

    @property
    def external_keyboard_connected(self) -> bool:
        return self.kb_monitor.connected

    def run(self, argv: list[str] | None = None) -> int:
        app = Gtk.Application(application_id=APP_ID)

        def on_startup(application):
            application.hold()
            try:
                self._setup_ui()
                self._setup_services()
            except Exception:
                log.exception("TouchFlow startup failed")
                raise
            if is_first_run():
                show_onboarding_if_needed(self._window)
            if self._virtual_keyboard or not self.config.behavior.startup_hidden:
                self.show_keyboard(manual=self._virtual_keyboard)

        app.connect("startup", on_startup)
        return app.run(argv if argv is not None else sys.argv)

    def _setup_ui(self) -> None:
        cfg = self.config

        self._window = Gtk.Window()
        self._window.set_title("TouchFlow")
        self._window.set_decorated(False)
        self._window.set_resizable(False)
        self._window.set_default_size(-1, cfg.layout.height_px)
        self._apply_keyboard_layer_shell()

        self._keyboard = KeyboardWidget(cfg, self.injector, self._on_keyboard_action)
        self._window.set_child(self._keyboard)

        if cfg.behavior.startup_hidden:
            self._window.set_visible(False)
        else:
            self._visible = True

        if cfg.behavior.swipe_from_bottom:
            self._swipe_zone = SwipeZone(
                cfg.behavior.swipe_zone_height_px,
                cfg.behavior.swipe_threshold_px,
                lambda: self.show_keyboard(manual=True),
            )
            self._swipe_zone.set_visible(True)

        if cfg.overlay.enabled:
            self._show_overlay()

    def _apply_keyboard_layer_shell(self) -> None:
        try:
            gi.require_version("Gtk4LayerShell", "1.0")
            from gi.repository import Gtk4LayerShell as LayerShell

            LayerShell.init_for_window(self._window)
            LayerShell.set_layer(self._window, LayerShell.Layer.OVERLAY)
            LayerShell.set_anchor(self._window, LayerShell.Edge.BOTTOM, True)
            LayerShell.set_anchor(self._window, LayerShell.Edge.LEFT, True)
            LayerShell.set_anchor(self._window, LayerShell.Edge.RIGHT, True)
            LayerShell.set_keyboard_mode(self._window, LayerShell.KeyboardMode.ON_DEMAND)
            if self.config.behavior.dock_bottom:
                LayerShell.set_exclusive_zone(self._window, self.config.layout.height_px)
        except (ImportError, ValueError):
            log.info("gtk4-layer-shell not found — using standard window placement")

    def _setup_services(self) -> None:
        self._dbus = TouchFlowDBusService(self)
        if not self._dbus.publish():
            log.error("D-Bus service failed to publish — CLI/settings won't work")
        self.focus_watcher.start()

        bindings = {
            "grab_device": self.config.bindings.grab_device,
            "toggle_visibility": self.config.bindings.toggle_visibility,
            "show_keyboard": self.config.bindings.show_keyboard,
            "hide_keyboard": self.config.bindings.hide_keyboard,
            "switch_layout": self.config.bindings.switch_layout,
            "toggle_overlay": self.config.bindings.toggle_overlay,
        }
        self._physical = PhysicalButtonListener(bindings, self._on_physical_action)
        self._physical.start()

        GLib.timeout_add_seconds(2, self._poll_external_kb)
        log.info("TouchFlow %s started", __version__)

    def _poll_external_kb(self) -> bool:
        self.kb_monitor.poll()
        return True

    def _on_external_kb_change(self, connected: bool) -> None:
        log.info("External keyboard %s", "connected" if connected else "disconnected")
        cfg = self.config.behavior
        if connected and cfg.hide_on_external_keyboard and self._visible:
            self.hide_keyboard()
        elif not connected and cfg.show_on_external_keyboard_disconnect and self._current_focus.is_text_entry:
            if not cfg.hide_on_external_keyboard or not connected:
                self.show_keyboard()

    def _on_focus_change(self, info: FocusInfo) -> None:
        self._current_focus = info
        if not info.is_text_entry:
            if self.config.behavior.auto_hide_on_blur and self._visible and not self._manual_show:
                self.hide_keyboard()
            return

        if info.app_id in self.config.excluded_apps:
            return
        if info.window_class in self.config.excluded_window_classes:
            return

        if self.config.behavior.hide_on_external_keyboard and self.kb_monitor.connected:
            return

        if not self.config.behavior.auto_show:
            return

        if not self.learning.should_auto_show(info.app_id, info.window_class):
            log.debug("Learning blocked auto-show for %s", info.app_id)
            return

        self.show_keyboard()
        self.learning.on_auto_show(info.app_id, info.window_class)

    def show_keyboard(self, manual: bool = False) -> None:
        if self.config.behavior.hide_on_external_keyboard and self.kb_monitor.connected:
            if not manual:
                return

        if self._window:
            self._window.set_visible(True)
            self._window.present()
            self._visible = True
            self._show_time = time.time()
            self._manual_show = manual
            if manual:
                self.learning.on_manual_show(
                    self._current_focus.app_id,
                    self._current_focus.window_class,
                )

    def hide_keyboard(self, manual: bool = False) -> None:
        if self._window:
            immediate = manual and (time.time() - self._show_time) < 1.5
            self.learning.on_dismiss(
                self._current_focus.app_id,
                self._current_focus.window_class,
                immediate=immediate,
            )
            self._window.set_visible(False)
            self._visible = False
            self._manual_show = False

    def toggle_keyboard(self) -> None:
        if self._visible:
            self.hide_keyboard(manual=True)
        else:
            self.show_keyboard(manual=True)

    def reload_config(self) -> None:
        self.config = load_config()
        self.learning.update_config(self.config.learning)
        if self._keyboard:
            self._keyboard.apply_config(self.config)
        if self._overlay:
            self._overlay.apply_config(self.config)
        log.info("Config reloaded")

    def reset_learning(self) -> None:
        self.learning.reset()
        log.info("Learning data reset")

    def toggle_overlay(self) -> None:
        if self._overlay:
            visible = self._overlay.get_visible()
            self._overlay.set_visible(not visible)
        else:
            self._show_overlay()

    def set_overlay_edit_mode(self, enabled: bool) -> None:
        if self._overlay:
            self._overlay.set_edit_mode(enabled)

    def _show_overlay(self) -> None:
        self._overlay = OverlayWindow(
            self.config,
            self.injector,
            on_config_change=lambda c: save_config(c),
        )
        self._overlay.set_visible(True)

    def _on_keyboard_action(self, action: str, detail: str = "") -> None:
        if action == "hide":
            self.hide_keyboard(manual=True)

    def _switch_language(self) -> None:
        if self._keyboard:
            self._keyboard._handle_action("switch_lang")

    def _on_physical_action(self, action: str) -> None:
        GLib.idle_add(self._dispatch_physical, action)

    def _dispatch_physical(self, action: str) -> bool:
        handlers = {
            "toggle": self.toggle_keyboard,
            "show_keyboard": lambda: self.show_keyboard(manual=True),
            "hide_keyboard": lambda: self.hide_keyboard(manual=True),
            "toggle_overlay": self.toggle_overlay,
            "switch_layout": self._switch_language,
        }
        fn = handlers.get(action)
        if fn:
            fn()
        return False


def main() -> None:
    import argparse

    from touchflow.dbus_client import try_show_existing_daemon

    parser = argparse.ArgumentParser(description="TouchFlow keyboard daemon")
    parser.add_argument(
        "--virtual-keyboard",
        action="store_true",
        help="Launched from KDE/GNOME virtual keyboard settings",
    )
    args, remaining = parser.parse_known_args()
    if args.virtual_keyboard:
        log.info("Virtual keyboard launch requested")
        if try_show_existing_daemon():
            log.info("Forwarded show to running daemon")
            return
        log.info("Starting daemon for virtual keyboard")
    daemon = TouchFlowDaemon(virtual_keyboard=args.virtual_keyboard)
    sys.exit(daemon.run(remaining if remaining else sys.argv))
